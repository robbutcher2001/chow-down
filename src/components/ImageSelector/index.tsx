import React, { Component, ChangeEvent } from 'react';

import styled from 'styled-components';
import questionImg from './question.svg';

const reader = new FileReader();

export interface ImageSelectorProps {
  name: string,
  label: string,
  form?: {
    [key: string]: string | number
  },
  setNewFormState?: (field: string, newValue: string | number) => void
};

interface OwnState {
  error: string
};

const Label = styled.label`
  display: flex;
  flex-direction: column;
  margin-bottom: 0.75rem;

  > input {
    border: solid 2px #e4e4e4;
  }

  > input {
    border-radius: 5px;
    padding: 0 0.5em;
    margin: 0.5rem 0;
    height: 2rem;
    font-size: 1rem;
  }

  > img {
    width: 100%;
    max-width: 400px;
    margin-bottom: 1rem;
    background-color: rgba(0,0,0,.2);
    cursor: pointer;
  }

  .red {
    color: #dc3545;
  }
`

class ImageSelector extends Component<ImageSelectorProps, OwnState> {
  constructor(props: ImageSelectorProps) {
    super(props);

    reader.onload = this.getImgData;

    this.state = {
      error: null
    };
  }

  componentDidMount = () => {
    this.props.setNewFormState(
      this.props.name,
      questionImg
    );
  };

  //TODO: select image, then open popup and click cancel, tries to render null
  getImgData = (event: ProgressEvent<FileReader>) => {
    this.props.setNewFormState(
      this.props.name,
      event.target.result.toString()
    );
    this.setState({
      error: null
    });
  };

  onChange = (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    const files: FileList = event.target.files;

    if (files.length === 1) {
      const img: File = files[0];
      if (img.type.match(/image.*/)) {
        reader.readAsDataURL(img);
      }
      else {
        this.props.setNewFormState(
          this.props.name,
          null
        );
        this.setState({
          error: 'Not an image'
        });
      }
    }
    else {
      this.props.setNewFormState(
        this.props.name,
        null
      );
      this.setState({
        error: 'Select only one image'
      });
    }
  };

  render = () => (
    <Label htmlFor={this.props.name}>
      {this.props.label}
      <input
        id={this.props.name}
        name={this.props.name}
        type='file'
        accept='image/*'
        onChange={event => this.onChange(event)} />
      <img src={this.props.form[this.props.name].toString()} />
      <div className='red'>{this.state.error}</div>
    </Label>
  );
};

export default ImageSelector;