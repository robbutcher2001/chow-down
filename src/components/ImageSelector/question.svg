<svg xml:space="preserve" width="300px" height="300px" viewBox="0 0 100 100" y="0" x="0" xmlns="http://www.w3.org/2000/svg" id="Layer_1" version="1.1" xmlns:xlink="http://www.w3.org/1999/xlink" style="width:100%;height:100%;background-size:initial;background-repeat-y:initial;background-repeat-x:initial;background-position-y:initial;background-position-x:initial;background-origin:initial;background-color:initial;background-clip:initial;background-attachment:initial">
  <path stroke="#333" fill="#666766" stroke-miterlimit="10" stroke-width="3.5" d="M41.258 66.254c-.019-.478-.047-.997-.077-1.545-.32-5.811-1.068-19.417 11.021-22.361 1.091-.266 2.063-.373 2.845-.46l.567-.063c3.978-.438 6.52-.776 8.553-2.167 1.729-1.182 2.961-2.876 3.381-4.651.382-1.609.209-3.347-.476-4.768-.29-.602-.839-1.31-2.883-2.414-3.204-1.728-7.375-2.461-13.131-2.322l-.344.009c-3.538.089-7.197.18-10.072.949-6.112 1.634-8.021 6.255-8.546 9.844l-14.333-2.097c1.586-10.834 8.74-18.961 19.139-21.741 4.535-1.212 9.271-1.331 13.449-1.436l.343-.008c8.387-.219 14.856 1.078 20.374 4.054 2.775 1.498 6.731 4.047 9.057 8.879 2.131 4.427 2.671 9.54 1.518 14.397-1.261 5.315-4.564 10.025-9.299 13.264-.47.322-.952.626-1.445.911-2.249 1.301-4.707 2.18-7.24 2.74-1.163.257-2.339.45-3.52.609-.795.107-1.606.048-2.348.399a3.801 3.801 0 0 0-1.656 1.561c-1.012 1.793-.599 4.116-.491 6.076.035.63.17 1.79.191 2.34H41.258z" style="stroke:rgb(51, 51, 51);fill:rgb(102, 103, 102)"></path>
  <path stroke="#333" fill="#666766" stroke-miterlimit="10" stroke-width="3.5" d="M58.657 80.236c0 4.84-3.496 8.764-9.562 8.764s-9.562-3.924-9.562-8.764 3.496-8.764 9.562-8.764 9.562 3.924 9.562 8.764z" style="stroke:rgb(51, 51, 51);fill:rgb(102, 103, 102)"></path>
  <path stroke-width="3.5" stroke="#000" stroke-miterlimit="10" stroke-linejoin="round" stroke-linecap="round" opacity=".2" d="M40.643 26.462c1.764-.472 3.825-.687 5.971-.804a32.979 32.979 0 0 0-4.394-.153l-.297.009c-3.054.089-6.212.18-8.693.949-4.847 1.501-6.628 5.524-7.237 8.951l6.104.893c.525-3.59 2.434-8.211 8.546-9.845z" style="stroke:rgb(0, 0, 0)"></path>
  <path stroke-width="3.5" stroke="#000" stroke-miterlimit="10" stroke-linejoin="round" stroke-linecap="round" opacity=".2" d="M80.126 23.956c-2.326-4.832-6.282-7.381-9.057-8.879-5.518-2.976-11.987-4.273-20.374-4.054l-.343.008c-1.108.028-2.257.058-3.429.108 4.897.401 8.968 1.688 12.568 3.938 2.395 1.498 5.81 4.047 7.817 8.879 1.839 4.427 2.305 9.54 1.31 14.397-1.088 5.315-3.939 10.025-8.026 13.264-.406.322-.822.626-1.248.911-1.941 1.301-4.063 2.18-6.248 2.74a32.33 32.33 0 0 1-3.038.609c-.686.107-1.386.048-2.026.399-.6.329-1.104.893-1.429 1.561-.873 1.793-.517 4.116-.424 6.076.03.63.146 1.79.165 2.34h9.493c-.022-.55-.156-1.711-.191-2.34-.109-1.96-.521-4.283.491-6.076a3.801 3.801 0 0 1 1.656-1.561c.742-.351 1.553-.292 2.348-.399a42.786 42.786 0 0 0 3.52-.609c2.532-.56 4.991-1.438 7.24-2.74.493-.285.975-.589 1.445-.911 4.736-3.239 8.039-7.949 9.299-13.264 1.151-4.857.612-9.97-1.519-14.397z" style="stroke:rgb(0, 0, 0)"></path>
  <path stroke-width="3.5" stroke="#000" stroke-miterlimit="10" stroke-linejoin="round" stroke-linecap="round" opacity=".2" d="M49.095 71.473a11.87 11.87 0 0 0-3.324.455c3.45 1.161 5.384 4.439 5.384 8.309 0 3.87-1.934 7.148-5.384 8.309 1.002.29 2.108.454 3.324.454 6.066 0 9.562-3.924 9.562-8.764s-3.496-8.763-9.562-8.763z" style="stroke:rgb(0, 0, 0)"></path>
  <circle fill="#fff" r="2.667" cy="75.74" cx="44.671" style="fill:rgb(255, 255, 255)"></circle>
  <path stroke-miterlimit="10" stroke-width="3.5" stroke-linecap="round" stroke="#fff" fill="none" d="M33.842 18.264c-.162.078-.323.158-.482.24" style="stroke:rgb(255, 255, 255)"></path>
  <path stroke-miterlimit="10" stroke-width="3.5" stroke-linecap="round" stroke="#fff" fill="none" d="M49.108 15.494c-3.573-.126-7.058.194-10.224.98" style="stroke:rgb(255, 255, 255)"></path>
</svg>