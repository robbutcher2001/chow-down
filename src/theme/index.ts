import { DefaultTheme } from 'styled-components';

import colour from './defs/_colour';

const theme: DefaultTheme = {
  colour
};

export default theme;