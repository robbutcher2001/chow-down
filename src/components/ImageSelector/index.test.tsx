import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from '../../themes';
import ImageSelector from '.';

const mockFormState = {
  title: 'value'
};

test('ImageSelector basic snapshot render', () => {
  const imageSelector = renderer.create(
    <ThemeProvider theme={theme}>
      <ImageSelector
        name='title'
        label='The Image Selector Title'
        validator={() => true}
        form={mockFormState}
        validFields={{}}
        setNewFormState={jest.fn}
        setValidationState={jest.fn}
      />
    </ThemeProvider>
  );

  expect(imageSelector.toJSON()).toMatchSnapshot();
});

test('ImageSelector correct tag content assertion', () => {
  const { getByText } = render(
    <ThemeProvider theme={theme}>
      <ImageSelector
        name='title'
        label='The Image Selector Title'
        validator={() => true}
        form={mockFormState}
        validFields={{}}
        setNewFormState={jest.fn}
        setValidationState={jest.fn}
      />
    </ThemeProvider>
  );

  expect(getByText(/The Image Selector Title/).textContent).toEqual('The Image Selector Title');
});