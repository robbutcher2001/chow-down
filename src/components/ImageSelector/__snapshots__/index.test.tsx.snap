// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSelector basic snapshot render 1`] = `
<label
  className="sc-AxirZ igNEsn"
  htmlFor="title"
>
  <span>
    The Image Selector Title
  </span>
  <input
    accept="image/jpeg,image/png"
    hidden={true}
    id="title"
    name="title"
    onChange={[Function]}
    type="file"
  />
  <figure>
    <img
      src="value"
    />
  </figure>
  <div />
</label>
`;
